File: app/src/main/java/net/squanchy/notification/NotificationService.java
package net.squanchy.notification;

import java.util.Collections;
import java.util.Comparator;
import java.util.List;

import net.squanchy.schedule.domain.view.Event;
import net.squanchy.service.firebase.FirebaseAuthService;
import net.squanchy.service.repository.EventRepository;

import io.reactivex.Observable;
import io.reactivex.schedulers.Schedulers;

import static net.squanchy.support.lang.Lists.filter;

class NotificationService {

    private final FirebaseAuthService authService;
    private final EventRepository eventRepository;

    NotificationService(FirebaseAuthService authService, EventRepository eventRepository) {
        this.authService = authService;
        this.eventRepository = eventRepository;
    }

    Observable<List<Event>> sortedFavourites() {
        return authService.ifUserSignedInThenObservableFrom(userId -> eventRepository.events(userId)
                .map(events -> filter(events, Event::getFavorited))
                .map(events -> {
                    Collections.sort(events, byStartDate());
                    return events;
                })
                .take(1)
                .subscribeOn(Schedulers.io()));
    }

    private Comparator<Event> byStartDate() {
        return (event1, event2) -> event1.getStartTime().compareTo(event2.getStartTime());
    }
}

File: app/src/main/java/net/squanchy/notification/NotificationService.kt
package net.squanchy.notification

import io.reactivex.Single
import io.reactivex.schedulers.Schedulers
import net.squanchy.schedule.domain.view.Event
import net.squanchy.service.repository.AuthService
import net.squanchy.service.repository.EventRepository

internal class NotificationService(private val authService: AuthService, private val eventRepository: EventRepository) {

    fun sortedFavourites(): Single<List<Event>> {
        return authService.ifUserSignedInThenObservableFrom { userId ->
            eventRepository.events(userId)
                .map { it.filter { it.favorite } }
                .map { it.sortedBy { it.startTime } }
                .subscribeOn(Schedulers.io())
        }.firstOrError()
    }
}
